File: problems/leetcode/0000/39/39.cpp
// 39. Combination Sum
// https://leetcode.com/problems/combination-sum/

using namespace std;

#include <vector>

class Solution {
public:
  vector<vector<int>> combinationSum(vector<int> &candidates, int target) {
    vector<vector<int>> answer;
    int candidateCount = candidates.size(), remainingTarget = target;
    vector<int> currentCombination;

    backtracking(answer, candidates, currentCombination, 0, target);

    return answer;
  }

  void backtracking(vector<vector<int>> &answer, vector<int> &candidates,
                    vector<int> &currentCombination, int index,
                    int remainingTarget) {
    if (candidates.size() <= index) {
      return;
    }

    int target = remainingTarget, candidate = candidates[index];

    while (target > 0) {
      backtracking(answer, candidates, currentCombination, index + 1, target);
      currentCombination.push_back(candidate);
      target -= candidate;
    }

    if (target == 0) {
      vector<int> possibleCombination(currentCombination);
      answer.push_back(possibleCombination);
    }

    while (target < remainingTarget) {
      currentCombination.pop_back();
      target += candidate;
    }
  }
};

File: problems/leetcode/0000/39/39.py
# 39. Combination Sum
# https://leetcode.com/problems/combination-sum/description/

# 1. Naive Solution
# 각각의 후보 숫자들을 고를 수 있는 경우의 수가 정해짐
# 각 경우에 따라 후속 숫자들을 고를 수 있는 경우의 수가 또 정해짐
# 반복해서 Backtracking으로 모든 경우의 수를 확인

from typing import *

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.combinations = []
        self.currentCombination = []

        self.traverseCases(candidates, 0, target)

        return self.combinations

    def traverseCases(self, candidates: List[int], index: int, target: int) -> None:
        if target == 0:
            self.combinations.append(list(self.currentCombination))
            return

        if index == len(candidates):
            return

        self.traverseCases(candidates, index + 1, target)
        for i in range(target // candidates[index]):
            self.currentCombination.append(candidates[index])
            self.traverseCases(candidates, index + 1, target - (i + 1) * candidates[index])

        for _ in range(target // candidates[index]):
            self.currentCombination.pop()
